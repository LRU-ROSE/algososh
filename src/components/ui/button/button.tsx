import React, { useMemo } from "react";
import styles from "./button.module.css";
import loaderIcon from "../../../images/icons/loader.svg";
import { AscendingIcon } from "../icons/ascending-icon";
import { DescendingIcon } from "../icons/descending-icon";
import { Direction } from "../../../types/direction";
import cx from "../../../helpers/cx";

export const enum ErrorDisable {
  None,
  FieldsetError,
  FormError,
}

interface ButtonProps extends React.HTMLProps<HTMLButtonElement> {
  text?: string;
  type?: "button" | "submit" | "reset";
  sorting?: Direction;
  linkedList?: "small" | "big";
  isLoader?: boolean;
  extraClass?: string;
  errorDisable?: ErrorDisable;
  testId?: string;
}

export const Button: React.FC<ButtonProps> = ({
  text,
  extraClass = "",
  type = "button",
  isLoader = false,
  sorting,
  errorDisable = ErrorDisable.FieldsetError,
  linkedList,
  disabled,
  testId,
  ...rest
}) => {
  const currentIcon =
    sorting === Direction.Ascending ? <AscendingIcon /> : <DescendingIcon />;
  const className = useMemo(() => {
    let errorClass;
    switch (errorDisable) {
      case ErrorDisable.None:
        errorClass = null;
        break;
      case ErrorDisable.FieldsetError:
        errorClass = styles.buttonDisableFieldsetError;
        break;
      case ErrorDisable.FormError:
        errorClass = styles.buttonDisableFormError;
        break;
    }
    return cx(
      "text text_type_button text_color_primary",
      styles.button,
      linkedList && styles[linkedList],
      isLoader && styles.loader,
      errorClass,
      extraClass
    );
  }, [linkedList, isLoader, errorDisable, extraClass]);

  return (
    <button
      className={className}
      type={type}
      disabled={isLoader || disabled}
      data-testid={testId}
      {...rest}
    >
      {isLoader ? (
        <img className={styles.loader_icon} src={loaderIcon} alt="Загрузка." />
      ) : (
        <>
          {sorting && currentIcon}
          <p className={cx('text', sorting && "ml-5")}>{text}</p>
        </>
      )}
    </button>
  );
};
