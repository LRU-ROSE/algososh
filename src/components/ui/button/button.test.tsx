import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react";

import { Button } from "./button";

describe("Тестирование компонента Button", () => {
  it("отрисовка кнопки с текстом", () => {
    const tree = renderer.create(<Button text="текст" />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("отрисовка кнопки без текста", () => {
    const tree = renderer.create(<Button />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("отрисовка заблокированной кнопки", () => {
    const tree = renderer
      .create(<Button disabled={true} text="текст" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("отрисовка кнопки с индикацией загрузки", () => {
    const tree = renderer.create(<Button isLoader />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("корректность вызова колбека при клике на кнопку", () => {
    const random = `test-button-${Math.random()}`;
    const callback = jest.fn();

    render(<Button text={random} onClick={callback} />);

    fireEvent.click(screen.getByText(random));

    expect(callback).toHaveBeenCalledTimes(1);
  });
});
