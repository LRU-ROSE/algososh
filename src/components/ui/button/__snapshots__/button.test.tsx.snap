// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестирование компонента Button отрисовка заблокированной кнопки 1`] = `
<button
  className="text text_type_button text_color_primary button buttonDisableFieldsetError"
  disabled={true}
  type="button"
>
  <p
    className="text"
  >
    текст
  </p>
</button>
`;

exports[`Тестирование компонента Button отрисовка кнопки без текста 1`] = `
<button
  className="text text_type_button text_color_primary button buttonDisableFieldsetError"
  type="button"
>
  <p
    className="text"
  />
</button>
`;

exports[`Тестирование компонента Button отрисовка кнопки с индикацией загрузки 1`] = `
<button
  className="text text_type_button text_color_primary button loader buttonDisableFieldsetError"
  disabled={true}
  type="button"
>
  <img
    alt="Загрузка."
    className="loader_icon"
    src="loader.svg"
  />
</button>
`;

exports[`Тестирование компонента Button отрисовка кнопки с текстом 1`] = `
<button
  className="text text_type_button text_color_primary button buttonDisableFieldsetError"
  type="button"
>
  <p
    className="text"
  >
    текст
  </p>
</button>
`;
